### PR TITLE
WIP: Refactor/change root dir resolution

### PR DESCRIPTION
**Pull request by tswicegood**
*TODO*

### CODE REVIEW

File: conda/config.py [hunk 1]
```diff
@@ -112,11 +112,54 @@ def load_condarc(path):
...
+
+    possible_default_paths = possible_root_dirs()
+    for possible in possible_default_paths:
+        if os.path.exists(possible):
+            return possible
+
+    # all else failed, fall back to the sys.prefix -- this should be
+    # deprecated with a proper warning but is included here for non-
+    # standard installation paths.
+    return sys.prefix if compatibility else possible_default_paths[0]
```
Comment (pelson) [inline conda/config.py]: sys.prefix seems like a reasonable place to look IMHO. Another reasonable place might be a few directories up from the location of ``import conda; print(conda.__file__)``.

Comment (tswicegood) [inline conda/config.py]: It's reasonable, but not particularly scalable.  It means that you have to execute the `conda` command to accurately determine where the root is.  Moving to a convention, such as this, means that there's a documented API for finding the package cache, environments directory, and so on that can be implemented in other languages.  Currently, any other implementation is required to execute at least `conda info --root` to determine where to start.

File: conda/config.py [hunk 2]
```diff
@@ -112,11 +112,54 @@ def load_condarc(path):
...
+    if 'CONDA_ROOT' in os.environ:
+        return expanded_user_path(os.environ['CONDA_ROOT'])
+
+    rc_root_dir = rc.get('root_dir', False)
+    if rc_root_dir is not False:
+        return rc_root_dir
+
+    possible_default_paths = possible_root_dirs()
+    for possible in possible_default_paths:
+        if os.path.exists(possible):
```
Comment (tswicegood) [inline conda/config.py]: Note: this should check for the existence of a `conda-meta` directory or some other file that denotes a valid conda system is present.

**Comment by asmeurer** [comment]
I should have been clearer upfront about this, but I am -1 to this. I think you are underestimating the number of people who change the default install location or Anaconda/Miniconda.

**Comment by tswicegood** [comment]
@asmeurer if there are lot of them then I want them to be explicit about where it's installed.  Otherwise, I can't get at it from Launcher without using Python too.

**Comment by ilanschnell** [comment]
I guess we can close this.

**Comment by pelson** [comment]
@ilanschnell - is there an ongoing discussion about the need for a "root" environment? I essentially think the concept is redundant, and can see a few alternative paradigms which may be of interest. Perhaps the mailinglist is the right place to start the discussion, if it hasn't already been talked about in depth...

**Comment by ilanschnell** [comment]
From my perspective a root environment (where conda can be installed into) is needed, because otherwise things will get too confusing.  The decision what conda can be installed only the root environment was made very early on.  What other "alternative paradigms" are you thinking about?

**Comment by pelson** [comment]
> What other "alternative paradigms" are you thinking about?

Conda already works pretty well outside of the root environment (provided you install it manually, or create an environment with a leading underscore). The biggest problem which is encountered when doing this is that the ```pkgs``` and ```envs``` directories are assumed to be put in the installed conda's prefix - if that assumption were broken, and the directories were actually put a level below the installed conda's prefix, you end up with a structure:

```
${CONDA_INSTALL_LOCATION}/
                           pkgs/
                           envs/
                                conda_root/
                                           bin/conda
                                conda_py2k_root/
                                           bin/conda
                                my_py2_env/
                                           bin/python
                                my_py3_env/
                                           bin/python
                           bin/
                               conda -> (symlink to) ../envs/conda_root/bin/conda
                               activate -> (symlink to) ../envs/conda_root/bin/activate
                               deactivate -> (symlink to) ../envs/conda_root/bin/deactivate
```

There are some details to thrash out about how conda finds its subcommands (e.g. ``conda-build``) but I think pulling that from the conda executable's path (i.e. not the symlink's path) is perfectly attainable. This approach has the advantage of never needing to remove ``${CONDA_INSTALL_LOCATION}/bin`` from the PATH within the activate script, as it is a very lightweight directory containing the 3 key conda top level executables. The other advantage is that symlinks would not be necessary within each environment that point to the "root" conda - this also makes it possible to install a different conda into another environment, activate that environment, and be able to use that version of conda (essentially, conda is no longer special)

**Comment by ilanschnell** [comment]
The reason I closed this issue is because I don't see any real benefits of allowing conda outside the root environment, but maintaining it sounds very hard (in particular answering support questions).
If you want another conda on your system, just install Miniconda again.

**Pull request closed without merging**